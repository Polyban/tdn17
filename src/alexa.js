import createLogger from 'debug'
import merge from 'deepmerge'
import { wrapIn } from './util'

const debug = createLogger('alexa:sdk')

export function createRequestHandler(requestHandlers, selector) {
  return (event) => {
    const canonizedEvent = merge(
      {
        session: { attributes: {}, new: true },
        request: { type: 'LaunchRequest' }
      },
      event
    )
    const selectedHandler = selector(requestHandlers, canonizedEvent)

    let handlerName = '<Unknown handler name>'
    const match = /^function (\w+)/.exec(selectedHandler)
    if (match) {
      handlerName = match[0]
    } else if (selectedHandler.displayName) {
      handlerName = selectedHandler.displayName
    }

    debug(`Calling handler '${handlerName}'`)
    debug('Session: %s',
      JSON.stringify(
        { new: canonizedEvent.session.new, attributes: canonizedEvent.session.attributes }
      )
    )
    debug('Request: %s',
      JSON.stringify(
        {
          type: canonizedEvent.request.type,
          intent: canonizedEvent.request.intent,
          dialogState: canonizedEvent.request.dialogState
        }
      )
    )

    const result = selectedHandler(canonizedEvent)
    return typeof canonizedEvent.session.attributes.STATE !== 'undefined'
      ? merge(
        {
          sessionAttributes: {
            STATE: canonizedEvent.session.attributes.STATE
          }
        },
        result
      )
      : result
  }
}

function cardSimple(title, content) {
  return {
    type: 'Simple',
    title,
    content,
  }
}

function cardStandard(title, text, smallImageUrl, largeImageUrl) {
  return {
    type: 'Standard',
    title,
    text,
    image: {
      smallImageUrl,
      largeImageUrl,
    },
  }
}

function createUtterance(outputSpeech) {
  let type = 'PlainText'
  let output = Array.isArray(outputSpeech)
    ? outputSpeech.map(wrapIn('p')).join('')
    : outputSpeech
  if (/[<>]+/.exec(output)) {
    type = 'SSML'
    output = `<speak>${output}</speak>`
  }
  return {
    type,
    [type === 'PlainText' ? 'text' : 'ssml']: output,
  }
}

class AlexaResponse {
  constructor() {
    this.props = {
      outputSpeech: null,
      card: null,
      reprompt: {
        outputSpeech: null,
      },
      shouldEndSession: false,
      directives: []
    }
  }

  say(outputSpeech) {
    this.props.outputSpeech = createUtterance(outputSpeech)
    return this
  }

  reprompt(outputSpeech) {
    this.props.reprompt.outputSpeech = createUtterance(outputSpeech)
    return this
  }

  card(type = 'Simple', ...cardProps) {
    /* eslint prefer-rest-params: 0 */
    const cardObject = arguments[0]
    if (typeof cardObject === 'object') {
      // set card props directly
      this.props.card = cardObject
    } else {
      // use card creator function to create card from arguments
      this.props.card = {
        cardSimple,
        cardStandard,
      }[`card${type}`](...cardProps)
    }
    return this
  }

  end(shouldEndSession) {
    this.props.shouldEndSession = shouldEndSession
    return this
  }

  elicitSlot(slotToElicit) {
    this.props.directives = [
      {
        type: 'Dialog.ElicitSlot',
        slotToElicit
      }
    ]
    return this
  }

  valueOf() {
    const { outputSpeech, card, reprompt, shouldEndSession, directives } = this.props
    const response = {
      outputSpeech,
      shouldEndSession,
      directives
    }
    if (reprompt.outputSpeech !== null) {
      response.reprompt = reprompt
    }
    if (card !== null) {
      const cardProps = {
        ...card
      }
      /* eslint no-array-constructor: 0 */
      Array('content', 'text').forEach((key) => {
        if (cardProps[key] && Array.isArray(cardProps[key])) {
          cardProps[key] = cardProps[key].join('\n')
        }
      })
      response.card = cardProps
    }
    return response
  }
}

export function say(speechOutput) {
  return new AlexaResponse().say(speechOutput)
}

export function elicit(speechOutput, slotToElicit) {
  return new AlexaResponse().say(speechOutput).elicitSlot(slotToElicit)
}

function getCanonicalSlotResolution(slot) {
  if (!slot.resolutions || !slot.resolutions.resolutionsPerAuthority
    || !slot.resolutions.resolutionsPerAuthority.length) {
    return null
  }
  const resolution = slot.resolutions.resolutionsPerAuthority[0]
  if (resolution.status.code !== 'ER_SUCCESS_MATCH') {
    return null
  }
  return resolution.values[0].value
}

export function getSlot(slots, slotName) {
  if (!slots || !slots[slotName] || typeof slots[slotName].value === 'undefined') {
    return null
  }
  return getCanonicalSlotResolution(slots[slotName])
}

export function withState(handler, state) {
  return (...args) => {
    const result = handler(...args)
    return merge(
      result,
      { sessionAttributes: { STATE: state } }
    )
  }
}
