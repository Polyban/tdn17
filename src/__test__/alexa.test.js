import { say } from '../alexa'

describe('Alexa', () => {
  it('should build a response using say()', () => {
    const response = say('Test utterance')
    expect(response.valueOf()).toMatchSnapshot()
  })

  it('should build a response using say() in SSML', () => {
    const response = say('<p>Test utterance</p>')
    expect(response.valueOf()).toMatchSnapshot()
  })

  it('should reprompt()', () => {
    const response = say('Test utterance').reprompt('<p>Reprompt utterance</p>')
    expect(response.valueOf()).toMatchSnapshot()
  })

  it('should add Simple card()', () => {
    const response = say('Test utterance')
      .card('Simple', 'Card title', 'Card content')
    expect(response.valueOf()).toMatchSnapshot()
  })

  it('should add Standard card()', () => {
    const response = say('Test utterance')
      .card(
        'Standard',
        'Card title',
        'Card text',
        'http://small-image-url',
        'http://large-image-url'
      )
    expect(response.valueOf()).toMatchSnapshot()
  })

  it('should build a response using end()', () => {
    const response = say('Test utterance').end(false)
    expect(response.valueOf()).toMatchSnapshot()
  })
})
