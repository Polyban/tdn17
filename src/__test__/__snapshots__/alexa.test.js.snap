// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alexa should add Simple card() 1`] = `
Object {
  "card": Object {
    "content": "Card content",
    "title": "Card title",
    "type": "Standard",
  },
  "directives": Array [],
  "outputSpeech": Object {
    "text": "Test utterance",
    "type": "PlainText",
  },
  "shouldEndSession": false,
}
`;

exports[`Alexa should add Standard card() 1`] = `
Object {
  "card": Object {
    "image": Object {
      "largeImageUrl": "http://large-image-url",
      "smallImageUrl": "http://small-image-url",
    },
    "text": "Card text",
    "title": "Card title",
    "type": "Standard",
  },
  "directives": Array [],
  "outputSpeech": Object {
    "text": "Test utterance",
    "type": "PlainText",
  },
  "shouldEndSession": false,
}
`;

exports[`Alexa should build a response using end() 1`] = `
Object {
  "directives": Array [],
  "outputSpeech": Object {
    "text": "Test utterance",
    "type": "PlainText",
  },
  "shouldEndSession": false,
}
`;

exports[`Alexa should build a response using say() 1`] = `
Object {
  "directives": Array [],
  "outputSpeech": Object {
    "text": "Test utterance",
    "type": "PlainText",
  },
  "shouldEndSession": false,
}
`;

exports[`Alexa should build a response using say() in SSML 1`] = `
Object {
  "directives": Array [],
  "outputSpeech": Object {
    "ssml": "<speak><p>Test utterance</p></speak>",
    "type": "SSML",
  },
  "shouldEndSession": false,
}
`;

exports[`Alexa should reprompt() 1`] = `
Object {
  "directives": Array [],
  "outputSpeech": Object {
    "text": "Test utterance",
    "type": "PlainText",
  },
  "reprompt": Object {
    "outputSpeech": Object {
      "ssml": "<speak><p>Reprompt utterance</p></speak>",
      "type": "SSML",
    },
  },
  "shouldEndSession": false,
}
`;
